backend/downsample_test.py

```diff
@@ -0,0 +1,303 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+"""Test Module for downsample.py"""
+
+import os
+import random
+import tempfile
+
+import pytest
+
+from downsample import average_downsample
+from downsample import downsample
+from downsample import lttb_downsample
+from downsample import max_min_downsample
+from downsample import secondary_downsample
+from downsample import strategy_reducer
+from downsample import triangle_area
+from utils import convert_to_csv
+
+
+class TestDownsampleClass:
+    """Test cllass for downsample.py"""
+
+    @pytest.fixture
+    def records(self):
+        """Generates test data."""
+        return [
+            [1573149236256988, 100, 'PPX_ASYS'],
+            [1573149236257088, 100, 'PPX_ASYS'],
+            [1573149236257188, 300, 'PPX_ASYS'],
+            [1573149236257288, 100, 'PPX_ASYS'],
+            [1573149236257388, 100, 'PPX_ASYS'],
+            [1573149236257488, 100, 'PPX_ASYS'],
+            [1573149236257588, 100, 'PPX_ASYS'],
+            [1573149236257688, 100, 'PPX_ASYS'],
+            [1573149236257788, 5, 'PPX_ASYS'],
+            [1573149236257888, 100, 'PPX_ASYS']
+        ]
+
+    @pytest.fixture
+    def records_complex_case(self):
+        """Generates complex test data."""
+        start = 1573149236256988
+        records_complex_case = [[start, 100, 'SYS']]
+        for _ in range(500):
+            records_complex_case.append(
+                [
+                    records_complex_case[-1][0] + random.randint(1E4, 1E6),
+                    random.randint(0, 500),
+                    'SYS'
+                ])
+
+        last = -1
+        for record in records_complex_case:
+            assert record[0] >= last
+            last = record[0]
+
+        return records_complex_case
+
+    @pytest.mark.parametrize('max_records', [0, 5, 10, 20, 30, 40, 1000000])
+    def test_limit_of_records(self, records, max_records):
+        """Tests if data is within the threshold."""
+        limit = min(len(records), max_records)
+        assert len(max_min_downsample(
+            records, True, max_records)) <= limit
+        assert len(max_min_downsample(
+            records, False, max_records)) <= limit
+        assert len(average_downsample(
+            records, max_records)) <= limit
+        assert len(lttb_downsample(
+            records, max_records)) <= limit
+
+    @pytest.mark.parametrize("point1,point2,point3,area", [
+        ([0, 0], [0, 1], [1, 0], 0.5),
+        ([0, 0], [0, 10], [10, 0], 50),
+        ([0, 0], [40, 0], [400, 20], 400),
+        ([0, 0], [40, 0], [400, 0], 0),
+    ])
+    def test_triangle_area(self, point1, point2, point3, area):
+        """Tests on triangle area."""
+        assert abs(triangle_area(point1, point2, point3) - area) <= 0.01
+
+    def test_lttb_downsample(self, records):
```

Comment:
break this down to 7 methods or use pytest.mark.parametrize

same comments applied to the rest of the test methods.